### PR TITLE
Chi/en 16370/sub queries

### PR DESCRIPTION
**Pull request by chitang**
Partial sub-queries support to allow processing joins with derived views.

### CODE REVIEW

File: soql-analyzer/src/main/scala/com/socrata/soql/AnalysisSerializer.scala [hunk 1]
```diff
@@ -289,6 +290,7 @@ class AnalysisSerializer[C,T](serializeColumn: C => String, serializeType: T =>
       writeGrouped(isGrouped)
       writeDistinct(analysis.distinct)
       writeSelection(selection)
+      maybeWrite(from) { x => out.writeStringNoTag(x) }
```
Comment (rjmac) [inline soql-analyzer/src/main/scala/com/socrata/soql/AnalysisSerializer.scala]: You'll want to do the version-bumping dance here and in the deserializer

**Comment by rjmac** [comment]
It is not immediately obvious to me what will happen if labels are reused inside subqueries, e.g., "... join (select a_col from aaaa-aaaa join bbbb-bbbb as x where a_col=x.col) as x where my_col = x.a_col"

File: soql-analyzer/src/main/scala/com/socrata/soql/AnalysisSerializer.scala (hunk 1 above)
Comment (chitang) [inline soql-analyzer/src/main/scala/com/socrata/soql/AnalysisSerializer.scala]: Joins aren't really in active use.  Therefore we can avoid the dance.  We all ignored version difference when we added outer joins.

**Comment by chitang** [comment]
Do you mean serialization dictionary collision when you say labels reused inside subqueries?

**Comment by chitang** [comment]
You mean both aaaa-aaaa and bbbb-bbbb have the same column name?

**Comment by rjmac** [comment]
I mean the inner and outer queries both alias something to `x`

**Comment by chitang** [comment]
i will look.  But I think we don't need to handle nested joins yet.

File: soql-analyzer/src/main/scala/com/socrata/soql/SoQLAnalyzer.scala [hunk 2]
```diff
@@ -416,6 +418,57 @@ case class SoQLAnalysis[ColumnId, Type](isGrouped: Boolean,
...
+      orderBy = orderBy.map(_.map(_.mapColumnIds(t2ToF2(qColumnIdNewColumnIdWithJoinsMap))))
+    )
+  }
+
+  /**
+    * Change function from taking tuple2 to function taking two parameters.
+    */
+  private def t2ToF2[A, B, X](f: Tuple2[A, B] => X)(a: A, b: B): X = {
+    f(Tuple2(a, b))
+  }
```
Comment (rjmac) [inline soql-analyzer/src/main/scala/com/socrata/soql/SoQLAnalyzer.scala]: This is `Function.untupled`

**Comment by chitang** [comment]
I'll remove/replace it next time.  Thanks

**Comment by chitang** [comment]
I might have mis-read an email and mistakenly thought that you approved the new commit.

**Pull request merged**